src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1079 @@
+    case kTypeOnMeshPrefix:
+        config.mOnMeshPrefix.mPrefix = mPrefix;
+        config.mOnMeshPrefix.mStable = mStable;
+        config.mOnMeshPrefix.SetFromTlvFlags(mFlags);
+        SuccessOrExit(Get<Local>().AddOnMeshPrefix(config.mOnMeshPrefix));
+        break;
+
+    case kTypeExternalRoute:
+        config.mExternalRoute.mPrefix = mPrefix;
+        config.mExternalRoute.mStable = mStable;
+        config.mExternalRoute.SetFromTlvFlags(static_cast<uint8_t>(mFlags));
+        SuccessOrExit(Get<Local>().AddHasRoutePrefix(config.mExternalRoute));
+        break;
+    }
+
+    registerWithLeader = true;
+    SetState(kAdded);
+    Get<Publisher>().NotifyPrefixEntryChange(kEventEntryAdded, mPrefix);
+
+exit:
+    return registerWithLeader;
+}
+
+bool Publisher::PrefixEntry::Remove(State aNextState)
+{
+    // Remove the prefix entry to the network data.
+
+    bool registerWithLeader = false;
+
+    VerifyOrExit((GetState() == kAdded) || (GetState() == kRemoving));
+
+    switch (mType)
+    {
+    case kTypeOnMeshPrefix:
+        IgnoreError(Get<Local>().RemoveOnMeshPrefix(mPrefix));
+        break;
+
+    case kTypeExternalRoute:
+        IgnoreError(Get<Local>().RemoveHasRoutePrefix(mPrefix));
+        break;
+    }
+
+    registerWithLeader = true;
+    Get<Publisher>().NotifyPrefixEntryChange(kEventEntryRemoved, mPrefix);
+
+exit:
+    SetState(aNextState);
+    return registerWithLeader;
+}
+
+void Publisher::PrefixEntry::Process(void)
+{
+    // This method checks the entries currently present in Network Data
+    // based on which it then decides whether or not take action
+    // (add/remove or keep monitoring).
+
+    uint8_t numEntries          = 0;
+    uint8_t numPreferredEntries = 0;
+    uint8_t desiredNumEntries   = 0;
+
+    // Do not make any changes if device is not attached, and wait
+    // for role change event.
+    VerifyOrExit(Get<Mle::Mle>().IsAttached());
+
+    VerifyOrExit(GetState() != kNoEntry);
+
+    switch (mType)
+    {
+    case kTypeOnMeshPrefix:
+        CountOnMeshPrefixEntries(numEntries, numPreferredEntries);
+        desiredNumEntries = kDesiredNumOnMeshPrefix;
+        break;
+    case kTypeExternalRoute:
+        CountExternalRouteEntries(numEntries, numPreferredEntries);
+        desiredNumEntries = kDesiredNumExternalRoute;
+        break;
+    }
+
+    UpdateState(numEntries, numPreferredEntries, desiredNumEntries);
+
+exit:
+    return;
+}
+
+void Publisher::PrefixEntry::CountOnMeshPrefixEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const
+{
+    const PrefixTlv *      prefixTlv;
+    const BorderRouterTlv *brSubTlv;
+    int8_t                 preference             = BorderRouterEntry::PreferenceFromFlags(mFlags);
+    uint16_t               flagsWithoutPreference = BorderRouterEntry::FlagsWithoutPreference(mFlags);
+
+    prefixTlv = Get<Leader>().FindPrefix(mPrefix);
+    VerifyOrExit(prefixTlv != nullptr);
+
+    brSubTlv = prefixTlv->FindSubTlv<BorderRouterTlv>(mStable);
```

Comment:
```suggestion
    brSubTlv = prefixTlv->FindSubTlv<BorderRouterTlv>();
```

Should iterate all border router subtlvs and prefer stable over unstable in below `for()` loop?

Reply:
It is a good point @librasungirl.

I actually did consider this when writing the code but decided to go with the current (and simpler) model where we treat prefix entries marked as stable completely separate from entries not marked as stable. 

- If our entry itself is stable, then this is what we want as the correct behavior,
    - i.e., if we see the same prefix in network data but not marked as stable, we should not count it as a match (and should still add ours if there are not enough matching stable ones).
- If our entry is itself not marked as stable then we can consider an existing stable prefix entry as a match (I did not go for it since it adds complexity to the code we need iterate over all matching BR subTLvs).

I think practically the `Publisher` will be used to publish stable entry. We can cosndier limiting the API to check and allow only stable prefix entries.

------

A related note:  I also considered earlier on how to determine a match related to other flags:
- If we see an existing prefix entry that has all the same flags as our entry but it also has some additional flags that we don't have in our entry, should we count it as a match or not? 
- After some thoughts on this, I decided not to go for this and require an exact match of flags.
   - The challenge with allowing and checking additional flags is that we need to take into account the meaning of  each flag and their combinations (whether in all cases a prefix which has more flags than ours is equivalent). 
   - It adds complexity in the code for a not-so-practical situation.

Reply:
If prefixes are the same, stable Prefix and unstable Prefix from different Servers would share the same outer stable Prefix TLV in the network data with two  border router sub-tlvs inside - the one for the `unstable` server would have `stable` flag as false as below picture

<img width="821" alt="Screen Shot 2021-07-08 at 5 23 09 PM" src="https://user-images.githubusercontent.com/19217991/124897818-3dc02300-e011-11eb-829b-61ed616f05e3.png">

Adding stable comparison with border router subtlvs in the network data seems not too complex.

---
Agree with your thoughts on the other flags.

Reply:
Yes. We will have a single Prefix TLV but it can possibly have two BR sub-TLVs one stable and one not stable. And each sub-TLV can have a bunch of entries (which we want to iterate through and count matching ones).

- If our entry is stable, then we want to find the stable sub-TLV, so the current code `prefixTlv->FindSubTlv<BorderRouterTlv>(mStable);` is behavior we want.

- If our entry itself is not stable, the current code/model counts the matching entries which are also not stable. If we want to change this to count both stable and not stable matches, we need to add another enclosing loop to go through both BR sub-TLVs (stable and not stable sub-TLVs). 
   - This is not terribly complex to add but it is not a single line change. 
   - Note that we cannot just use the `FindSubTlv(void)` flavor instead of `FindSubTlv(mStable)`. This won't really iterate over both sub-TLVs.  `FindSubTlv(void)` will find the "first" matching TLV (ignoring the stable flag).

My goal here has been to avoid adding additional logic and behavior in the code to optimize the case that is not really intended as the main use (i.e., using `Publisher` with prefix entries that are not marked as stable). And if we change it since it would be a new behavior we probably need to also add test-case for it (again for something that is not intended to be used).

Considering more on this, I think  it may be better to change the `Publisher` public methods to only allow stable prefix entries. This help keep the code simple and make the intended use clear. I will change this in the next push.

Reply:
Not considering the added code complexity, which I believe is small, do we have good reason to not prefer stable to unstable?

Reply:
@simonlingoogle the existing code (from the initial implementation) already did prefer the stable over unstable matches (I explained in above comments under the case "if our entry is stable" and that's the behavior we do want).

The question was whether we want to have `Publisher` support publishing entries which are not stable and add complexity to the code to optimize how we handle them. This is not really a practical or intended use for `Publisher` so we now simply don't allow `Publisher` to be used with non-stable entries.

Reply:
Fine to me if we simply don't allow publishing non-stable entries.